src/main/java/play/dev/filewatch/SourceModificationWatch.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package play.dev.filewatch;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+public final class SourceModificationWatch {
+
+  private SourceModificationWatch() {}
+
+  private static Set<File> listFiles(Supplier<Iterable<File>> sourcesFinder) {
+    return StreamSupport.stream(sourcesFinder.get().spliterator(), false)
+        .collect(Collectors.toSet());
+  }
+
+  private static long findLastModifiedTime(File file) {
+    try {
+      return Files.getLastModifiedTime(file.toPath()).toInstant().toEpochMilli();
+    } catch (IOException e) {
+      return 0L;
+    }
+  }
+
+  private static long findLastModifiedTime(Set<File> files) {
+    if (!files.isEmpty()) {
+      return files.stream()
+          .max(Comparator.comparingLong(SourceModificationWatch::findLastModifiedTime))
+          .map(SourceModificationWatch::findLastModifiedTime)
+          .get();
+    } else {
+      return 0L;
+    }
+  }
+
+  public static class Result {
+    private final boolean triggered;
+
+    private final WatchState state;
+
+    public boolean isTriggered() {
+      return triggered;
+    }
+
+    public WatchState getState() {
+      return state;
+    }
+
+    public Result(boolean triggered, WatchState state) {
+      this.triggered = triggered;
+      this.state = state;
+    }
+  }
+
+  public static Result watch(
+      final Supplier<Iterable<File>> sourcesFinder,
+      final int pollDelayMillis,
+      WatchState state,
+      final Supplier<Boolean> terminationCondition) {
+    while (true) {
+      var filesToWatch = listFiles(sourcesFinder);
```

Comment:
You called `sourcesFinder.GET()` here already, but this should happen _within_ the `listFiles` method.